narwhal/worker/src/quorum_waiter.rs

```diff
@@ -106,36 +117,53 @@ impl QuorumWaiter {
+                        // if we terminate before we have managed to get to the full 2/3 stake.
+                        let _ = &wait_for_quorum;
+                        loop{
+                            if let Some(stake) = wait_for_quorum.next().await {
                                 total_stake += stake;
                                 if total_stake >= threshold {
-                                    if self.tx_batch.send(batch).await.is_err() {
-                                        tracing::debug!("{}", DagError::ShuttingDown);
-                                    }
-                                    break;
-                                }
-                            }
-
-                            result = self.rx_reconfigure.changed() => {
-                                result.expect("Committee channel dropped");
-                                let message = self.rx_reconfigure.borrow().clone();
-                                match message {
-                                    ReconfigureNotification::NewEpoch(new_committee)
-                                        | ReconfigureNotification::UpdateCommittee(new_committee) => {
-                                            self.network.cleanup(self.committee.network_diff(&new_committee));
-                                            self.committee = new_committee;
-                                            tracing::debug!("Dropping batch: committee updated to {}", self.committee);
-                                            break; // Don't wait for acknowledgements.
-                                    },
-                                    ReconfigureNotification::Shutdown => return
+                                    break
                                 }
+                            } else {
+                                tracing::error!("Batch dissemination ended without a quorum.");
+                                break;
                             }
                         }
+                        (batch, wait_for_quorum)
+                    });
+                },
+
+                // Process futures in the pipeline. They complete when we have sent to >2/3
+                // of other worker by stake, but after that we still try to send to the remaining
+                // on a best effort basis.
+                Some((batch, mut remaining)) = pipeline.next() => {
+
+                    // Attempt to send messages to the remaining workers
+                    best_effort_with_timeout.push(async move {
+                        // Bound the attempt to a few seconds to tolerate nodes that are
+                        // offline and will never succeed.
+                        //
+                        // TODO: make the constant a config parameter.
+                        timeout(Duration::from_secs(5), async move{
+                            while remaining.next().await.is_some() { }
+                        }).await
+                    });
```

Comment:
I really like the approach of giving the chance for more than 2/3 to get payload, I am tempted to though to say "let's make one change at a time". We are already optimising part of the process by broadcasting multiple batches at once. I would deploy first that , benchmark, observe the effect. Then do the second change and repeat. Too overcautious? Maybe, but given the current situation I would like to more clearly see the effects of the changes.